opendkim/opendkim-atpszone.8.in
.TH opendkim-atpszone 8 "The Trusted Domain Project"
.SH NAME
.B opendkim-atpszone
\- DKIM ATPS zone file generation tool
.SH SYNOPSIS
.B opendkim-atpszone
[\-A]
[\-C address]
[\-E secs]
[\-h hash]
[\-N ns[,...]]
[\-o file]
[\-r secs]
[\-R secs]
[\-S]
[\-t secs]
[\-T secs]
[\-u domain]
[\-v]
[dataset]
.SH DESCRIPTION
.B opendkim-atpszone
generates a file suitable for use with
.I named(8)
to publish a set of domains authorized as third-party signers for a local
domain.

The
.I dataset
parameter should specify a set of data as described in the
.I opendkim(8)
man page.  It can currently refer to flat files, Sleepycat databases,
comma-separated lists, LDAP directories or SQL databases.  The keys in the
named database are assumed to comprise a set of domains that are to be
advertised using the experimental Authorized Third-Party Signers protocol
as permitted to sign mail using DKIM on behalf of the local domain.  Values
in the database are not used.
.SH OPTIONS
.TP
.I \-A
Adds a "._atps" suffix to records in the zone file.
.TP
.I \-C contact
Uses
.I contact
as the contact information to be used when an SOA record is generated (see
.I \-S
below).  If not specified, the userid of the executing user and the local
hostname will be used; if the executing user can't be determined,
"hostmaster" will be used.
.TP
.I \-E secs
When generating an SOA record (see
.I \-S
below), use
.I secs
as the default record expiration time.  The default is 604800.
.TP
.I \-h hash
Specifies which SHA hash algorithm to use.  Must be one of "none", "sha1"
and "sha256", with "sha256" being the default if it is available.
.TP
.I \-N nslist
Specifies a comma-separated list of nameservers, which will be output in
NS records before the TXT records.  The first nameserver in this list will
also be used in the SOA record (if
.I \-S
is also specified) as the authority hostname.
.TP
.I \-o file
Sends output to the named
.I file
rather than standard output.
.TP
.I \-r secs
When generating an SOA record (see
.I \-S
below), use
.I secs
as the zone refresh time.  The default is 10800.
.TP
.I \-R secs
When generating an SOA record (see
.I \-S
below), use
.I secs
as the zone retry time.  The default is 1800.
.TP
.I \-S
Asks for an SOA record to be generated at the top of the output.  The
content of this output can be controlled using the
.I \-E,
.I \-r,
.I \-R,
.I \-T
options.  The serial number will be generated based on the current time of
day.
.TP
.I \-t ttl
Puts a TTL (time-to-live) value of
.I ttl
on all records output.  The units are in seconds.
.TP
.I \-T secs
When generating an SOA record (see
.I \-S
below), use
.I secs
as the default record TTL time.  The default is 86400.
.TP
.I \-u domain
Produce output suitable for use as input to
.B nsupdate(8)
to add ATPS records to the named
.I domain.
.TP
.I \-v
Increases the verbosity of debugging output written to standard error.
.SH VERSION
This man page covers the version of
.I opendkim-atpszone
that shipped with version @VERSION@ of
.I OpenDKIM.
.SH COPYRIGHT
Copyright (c) 2011, 2012, The Trusted Domain Project.  All rights reserved.
.SH SEE ALSO
.I nsupdate(8),
.I opendkim(8),
.I opendkim.conf(5)


RELEASE_NOTES
			OPENDKIM RELEASE NOTES

This listing shows the versions of the OpenDKIM package, the date of
release, and a summary of the changes in that release.

2.11.0		2018/11/??
	Add support for ED25519 keys.  GNUTLS support provided by
		Dilyan Palauzov.
	Add "header.s" and "header.a" in generated Authentication-Results
		header fields.
	Add _FFR_CONDITIONAL, experimental support for conditional DKIM
		signatures.
	Make odkim.internal_ip() available to all Lua hooks.  Problem noted
		by Iosif Fettich.
	Make bind code DNSSEC-aware.  Patch from Jack Bates.
	Extend KeyTable to specify signing algorithm.
		Patch from Yasuhito Futatsuki.
	Fix dkimf_db_nextpunct() so it doesn't incorrectly identify an encoded
		hex digit as a value delimiter.
	Fix issue #8: The password file critical section isn't big enough.
		Patch from Dilyan Palauzov.
	Fix issue #9: Plug a few failure mode memory leaks.  Patch from
		Dilyan Palauzov.
	Fix issue #15: Don't automatically skip the body when one mode (sign
		or verify) doesn't need it, because the other one might.
		Patch from Dilyan Palauzov.
	Fix issue #28: RequiredHeaders should report and log the specific
		error and reject the message, as documented.  Patch from
		Dilyan Palauzov.
	Fix issue #222: When the "Minimum" option is set, don't treat
		empty body as partial. Patch from Yasuhito FUTATSUKI
		(PR #223).
	Fix bug #234: Make "NoHeaderB" do something.  Reported by Joseph
		Coffland.
	Fix bug #235: Quote "header.b" values in case they contain a slash
		(which must be quoted).
	Fix bug #237: Fix processing of "SoftStart".
	CONFIG: Add compatibility with openssl-1.1.0.  Patch from
		Scott Kitterman.
	CONTRIB: Simplfy some logic in contrib/repute.  Patch from
		Dilyan Palauzov.
	LIBOPENDKIM: Feature request #190: Reject signature object requests
		where the domain name or selector includes non-printable
		characters.  Suggested by Franck Martin.
	LIBOPENDKIM: Re-fix bug #226: Deal with header fields that are
		wrapped before there's any content.  Somehow the wrong fix
		was committed.  Originally reported by Alessandro Vesely;
		re-reported by David Stevenson.
	LIBOPENDKIM: Fix bug #233: In the weird case of duplicate signatures,
		constrain the size of the resulting "header.b" value.  Problem
		noted by Joseph Coffland.
	LIBOPENDKIM: Fix bug #270: Don't set an upper bound on the size of
		the canonicalization buffer, which interferes with processing
		oversized input lines.  Reported by Mars Peng.
	LIBOPENDKIM: Fix parsing bug in dkim_mail_parse_multi(), where quotes
		were not being properly handled.
	LIBOPENDKIM: Expose conversion table between internal code already
		provided as DKIM_ macros and their literal name in C string.
	LIBOPENDKIM: Extend dkim_test_key() to allow testing a ed25519 key,
		as dkim_test_key2(). For API compatibility, function interface
		of dkim_test_key() is not changed, but it only calls
		dkim_test_key2(). Patch from Yasuhito Futatsuki.
	TOOLS: Feature requrest #187: Add option to match subdomains when
		generating zone files.  Patch from Andreas Schulze.
	TOOLS: issue #183: On opendkim-testkey, add support for ed25519 keys.
		Patch from Yasuhito Futatsuki.

2.10.3		2015/05/12
	LIBOPENDKIM: Make strict header checking non-destructive.  The last
		change to this code resulted in failing signatures.  Reported
		by Pedro Morales and Wez Furlong.

2.10.2		2015/05/10
	Fix bug #221: Report a DKIM result of "policy" if MinimumKeyBits
		or UnprotectedKey cause the signature to result in a "pass"
		override.  Reported by Kurt Roeckx.
	Fix bug #227: Revert removal of SenderHeaders configuration setting.
		Document that it is now limited to signature selection.
	LIBOPENDKIM: Fix bug #226: Deal with header fields that are
		wrapped before there's any content.  Reported by
		Alessandro Vesely.
	CONTRIB: Update to contrib/systemd/opendkim.service.in from
		Steve Jenkins.

2.10.1		2015/02/03
	Fix bug #214: Handle arbitrarily large From: fields.  Reported by
		Tomohiko Sasaki.
	Fix bug #220: Make DB_SIGNINGTABLE symbol available in Lua scripts.
		Problem noted by Klaus Heinrich.
	LIBOPENDKIM: Fix bug #213: Remove "dkim_default_senderhdrs" from
		dkim.h.  Problem noted by Daniel J. Luke.
	LIBOPENDKIM: Fix bug #219: Unresolved CNAMEs are not failures,
		according to the DNS (see RFC6604), so report them as
		NXDOMAIN or similar.  Reported by Alessandro Vesely.

2.10.0		2014/12/27
	Feature request #182: Remove "AddAllSignatureResults".  All signature
		results will now be added via Authentication-Results header
		fields.  Requested by Tomki Camp.
	Feature request #180: Rename "LDAPSoftStart" to "SoftStart" and apply
		it to SQL connections as well.  Requested by Daniel Kauffman.
	Feature request #179: Add "IgnoreMalformedMail" option.
	Fix bug #183: Discontinue support for ADSP.  This removes the
		following configuration file items:
		AddAllSignatureResults	LocalADSP
		ADSPAction		NoDiscardableMailTo
		ADSPNoSuchDomain	On-PolicyError
		BogusPolicy		SendADSPReports
		DisableADSP		SenderHeaders
		LDAPSoftStart		UnprotectedPolicy
	Make "rrvs" and "smime" recognized Authentication-Results methods.
	LIBOPENDKIM: Feature request #157: Add dkim_mail_parse_multi().
		Suggested by Alessandro Vesely.
	LIBOPENDKIM: Feature request #185: Add dkim_set_dnssec().  Patch
		from Alec Peterson.
	LIBOPENDKIM: Fix bug #183: Discontinue support for ADSP.  This
		means all of the following:
		- the dkim_policy_t type has been removed
		- the DKIM_POLICY_* constants have been removed
		- the DKIM_PRESULT_* constants have been removed
		- passing DKIM_OPTS_SENDERHDRS to dkim_options() now
		  results in an error
		- the DKIM_PSTATE structure has been removed
		- all of the following functions have been removed:
		  dkim_policy(), dkim_policy_dnssec(),
		  dkim_policy_getqueries(), dkim_policy_getreportinfo(),
		  dkim_policy_state_free(), dkim_policy_state_new(),
		  dkim_policy_syntax(), dkim_getpolicystr(),
		  dkim_getpresult(), dkim_getpresultstr(),

opendkim/opendkim-atpszone.c
/*
**  Copyright (c) 2010-2012, The Trusted Domain Project.  All rights reserved.
**
*/

#include "build-config.h"

/* system includes */
#include <sys/param.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <sysexits.h>
#include <string.h>
#include <assert.h>
#include <fcntl.h>
#include <ctype.h>
#include <stdlib.h>
#include <stdio.h>
#include <errno.h>
#include <unistd.h>
#include <pwd.h>

/* openssl includes */
#ifdef USE_GNUTLS
# include <gnutls/gnutls.h>
# include <gnutls/crypto.h>
# include <gnutls/abstract.h>
# include <gnutls/x509.h>
#else /* USE_GNUTLS */
# include <openssl/rsa.h>
# include <openssl/pem.h>
# include <openssl/evp.h>
# include <openssl/bio.h>
#endif /* USE_GNUTLS */

#ifndef FALSE
# define FALSE		0
#endif /* ! FALSE */
#ifndef TRUE
# define TRUE		1
#endif /* ! TRUE */
#ifndef SHA_DIGEST_LENGTH
# define SHA_DIGEST_LENGTH 20
#endif /* ! SHA_DIGEST_LENGTH */
#ifndef SHA256_DIGEST_LENGTH
# define SHA256_DIGEST_LENGTH 32
#endif /* ! SHA256_DIGEST_LENGTH */

/* libopendkim includes */
#include <dkim.h>

/* opendkim includes */
#include "opendkim-db.h"
#include "util.h"
#include "config.h"
#include "opendkim-config.h"

/* definitions */
#define	ATPSZONE	"._atps"
#define	BASE32_LENGTH	32
#define	BUFRSZ		256
#define	CMDLINEOPTS	"AC:E:h:o:N:r:R:St:T:u:v"
#define	DEFEXPIRE	604800
#define	DEFREFRESH	10800
#define	DEFRETRY	1800
#define	DEFTTL		86400
#define	HOSTMASTER	"hostmaster"
#define	MAXNS		16
#define	VALIDATPS	"v=ATPS1"

/* globals */
char *progname;

/*
**  USAGE -- print usage message and exit
**
**  Parameters:
**  	None.
**
**  Return value:
**  	EX_USAGE
*/

int
usage(void)
{
	fprintf(stderr, "%s: usage: %s [opts] [dataset]\n"
	                "\t-A          \tinclude '._atps' suffix\n"
	                "\t-C user@host\tcontact address to include in SOA\n"
	                "\t-E secs     \tuse specified expiration time in SOA\n"
	                "\t-h hash     \thash algorithm\n"
	                "\t-o file     \toutput file\n"
	                "\t-N ns[,...] \tlist NS records\n"
	                "\t-r secs     \tuse specified refresh time in SOA\n"
	                "\t-R secs     \tuse specified retry time in SOA\n"
	                "\t-S          \twrite an SOA record\n"
	                "\t-t secs     \tuse specified per-record TTL\n"
	                "\t-T secs     \tuse specified default TTL in SOA\n"
	                "\t-u domain   \tproduce nsupdate(8) script\n"
	                "\t-v          \tverbose output\n",
		progname, progname);

	return EX_USAGE;
}

/*
**  MAIN -- program mainline
**
**  Parameters:
**  	argc, argv -- the usual
**
**  Return value:
**  	Exit status.
*/

int
main(int argc, char **argv)
{
	_Bool suffix = FALSE;
	_Bool writesoa = FALSE;
	_Bool nsupdate = FALSE;
	int c;
	int status;
	int verbose = 0;
	int ttl = -1;
	int defttl = DEFTTL;
	int expire = DEFEXPIRE;
	int refresh = DEFREFRESH;
	int retry = DEFRETRY;
	int nscount = 0;
	time_t now;
	size_t dlen;
	size_t b32len;
	size_t shalen;
	char *p;
	char *dataset = NULL;
	char *udomain = NULL;
	char *outfile = NULL;
	char *hash = NULL;
	char *contact = NULL;
	char *nameservers = NULL;
	char *nslist[MAXNS];
	FILE *out;
	DKIMF_DB db;
#ifdef USE_GNUTLS
	gnutls_hash_hd_t sha;
#endif /* USE_GNUTLS */
	char domain[DKIM_MAXHOSTNAMELEN + 1];
	char hostname[DKIM_MAXHOSTNAMELEN + 1];
#ifdef HAVE_SHA256
	char shaout[SHA256_DIGEST_LENGTH];
#else /* HAVE_SHA256 */
	char shaout[SHA_DIGEST_LENGTH];
#endif /* HAVE_SHA256 */
	char base32[BASE32_LENGTH + 1];
	char value[BUFRSZ + 1];

	progname = (p = strrchr(argv[0], '/')) == NULL ? argv[0] : p + 1;

	while ((c = getopt(argc, argv, CMDLINEOPTS)) != -1)
	{
		switch (c)
		{
		  case 'A':
			suffix = TRUE;
			break;

		  case 'C':
			contact = strdup(optarg);
			break;

		  case 'E':
			expire = strtol(optarg, &p, 10);
			if (*p != '\0' || expire < 0)
			{
				fprintf(stderr, "%s: invalid expire value\n",
				        progname);
				return EX_USAGE;
			}
			break;

		  case 'h':
			if (strcasecmp(optarg, "sha1") != 0 &&
			    strcasecmp(optarg, "sha256") != 0 &&
			    strcasecmp(optarg, "none") != 0)
			{
				fprintf(stderr, "%s: invalid hash algorithm\n",
				        progname);
				return EX_USAGE;
			}
#ifndef HAVE_SHA256
			else if (strcasecmp(optarg, "sha256") == 0)
			{
				fprintf(stderr,
				        "%s: hash algorithm \"%s\" not supported\n",
				        progname, optarg);
				return EX_SOFTWARE;
			}
#endif /* ! HAVE_SHA256 */
			hash = optarg;
			if (strcasecmp(hash, "sha1") == 0)
				shalen = SHA_DIGEST_LENGTH;
			else
				shalen = SHA256_DIGEST_LENGTH;
			break;

		  case 'N':
			nameservers = strdup(optarg);
			break;

		  case 'o':
			outfile = optarg;
			break;

		  case 'r':
			refresh = strtol(optarg, &p, 10);
			if (*p != '\0' || refresh < 0)
			{
				fprintf(stderr, "%s: invalid refresh value\n",
				        progname);
				return EX_USAGE;
			}
			break;

		  case 'R':
			retry = strtol(optarg, &p, 10);
			if (*p != '\0' || retry < 0)
			{
				fprintf(stderr, "%s: invalid retry value\n",
				        progname);
				return EX_USAGE;
			}
			break;

		  case 'S':
			writesoa = TRUE;
			break;

		  case 't':
			ttl = strtol(optarg, &p, 10);
			if (*p != '\0' || ttl < 0)
			{
				fprintf(stderr, "%s: invalid TTL value\n",
				        progname);
				return EX_USAGE;
			}
			break;

		  case 'T':
			defttl = strtol(optarg, &p, 10);
			if (*p != '\0' || defttl < 0)
			{
				fprintf(stderr,
				        "%s: invalid default TTL value\n",
				        progname);
				return EX_USAGE;
			}
			break;

		  case 'u':
			nsupdate = TRUE;
			udomain = optarg;
			break;

		  case 'v':
			verbose++;
			break;

		  default:
			return usage();
		}
	}

	if (optind != argc)
		dataset = argv[optind];

	status = dkimf_db_open(&db, dataset, DKIMF_DB_FLAG_READONLY,
	                       NULL, NULL);
	if (status != 0)
	{
		fprintf(stderr, "%s: dkimf_db_open() failed\n", progname);
		return 1;
	}

	if (dkimf_db_type(db) == DKIMF_DB_TYPE_REFILE)
	{
		fprintf(stderr, "%s: invalid data set type\n", progname);
		(void) dkimf_db_close(db);
		return 1;
	}

	if (verbose > 0)
		fprintf(stderr, "%s: database opened\n", progname);

	if (outfile != NULL)
	{
		out = fopen(outfile, "w");
		if (out == NULL)
		{
			fprintf(stderr, "%s: %s: fopen(): %s\n",
			        progname, outfile, strerror(errno));
			(void) dkimf_db_close(db);
			return 1;
		}
	}
	else
	{
		out = stdout;
	}

	if (nameservers != NULL)
	{
		for (p = strtok(nameservers, ",");
		     p != NULL && nscount < MAXNS;
		     p = strtok(NULL, ","))
			nslist[nscount++] = p;
	}

	memset(hostname, '\0', sizeof hostname);
	gethostname(hostname, sizeof hostname);

	if (nscount == 0)
		nslist[nscount++] = hostname;

	(void) time(&now);

	if (!nsupdate)
	{
		fprintf(out, "; DKIM ATPS zone data\n");
		fprintf(out, "; auto-generated by %s at %s\n", progname,
		        ctime(&now));
	}

	if (writesoa && !nsupdate)
	{
		struct tm *tm;

		fprintf(out, "@\tIN\tSOA\t%s\t", nslist[0]);

		if (contact != NULL)
		{
			for (p = contact; *p != '\0'; p++)
			{
				if (*p == '@')
					*p = '.';
			}

			fprintf(out, "%s", contact);
		}
		else
		{
			struct passwd *pwd;

			pwd = getpwuid(getuid());

			fprintf(out, "%s.%s",
			        pwd == NULL ? HOSTMASTER : pwd->pw_name,
			        hostname);
		}

		tm = localtime(&now);

		fprintf(out,
		        "\t (\n"
		        "\t%04d%02d%02d%02d   ; Serial (yyyymmddhh)\n"
		        "\t%-10d   ; Refresh\n"
		        "\t%-10d   ; Retry\n"
		        "\t%-10d   ; Expire\n"
		        "\t%-10d ) ; Default\n\n",
		        tm->tm_year + 1900,
		        tm->tm_mon + 1,
		        tm->tm_mday,
		        tm->tm_hour,
		        refresh, retry, expire, defttl);
	}

	if (nameservers != NULL && !nsupdate)
	{
		for (c = 0; c < nscount; c++)
			fprintf(out, "\tIN\tNS\t%s\n", nslist[c]);

		fprintf(out, "\n");
	}

	if (nsupdate)
	{
		fprintf(out, "server %s\n", nslist[0]);
		fprintf(out, "zone %s\n", udomain);
	}

#ifdef USE_GNUTLS
	(void) gnutls_global_init();
#endif /* USE_GNUTLS */

	for (c = 0; ; c++)
	{
		memset(domain, '\0', sizeof domain);
		dlen = sizeof domain;

		status = dkimf_db_walk(db, c == 0, domain, &dlen, NULL, 0);
		if (status == -1)
		{
			fprintf(stderr, "%s: dkimf_db_walk(%d) failed\n",
			        progname, c);
			(void) dkimf_db_close(db);
			return 1;
		}
		else if (status == 1)
		{
			break;
		}

		/* convert to lowercase */
		dkimf_lowercase(domain);

		memset(base32, '\0', sizeof base32);

		if (hash == NULL || strcasecmp(hash, "none") != 0)
		{
			/* compute SHA1 hash */
#ifdef USE_GNUTLS
# ifdef HAVE_SHA256
			if (hash == NULL || strcasecmp(hash, "sha256") == 0)
			{
				(void) gnutls_hash_init(&sha,
				                        GNUTLS_DIG_SHA256);
			}
			else
			{
				(void) gnutls_hash_init(&sha, GNUTLS_DIG_SHA1);
			}
# else /* HAVE_SHA256 */
			(void) gnutls_hash_init(&sha, GNUTLS_DIG_SHA1);
# endif /* HAVE_SHA256 */
			(void) gnutls_hash(sha, domain, strlen(domain));
			(void) gnutls_hash_deinit(sha, shaout);
#else /* USE_GNUTLS */
# ifdef HAVE_SHA256
			if (hash == NULL || strcasecmp(hash, "sha256") == 0)
			{
				(void) EVP_Digest(domain, strlen(domain),
				                  shaout, NULL, EVP_sha256(),
				                  NULL);
			}
			else
			{
				(void) EVP_Digest(domain, strlen(domain),
				                  shaout, NULL, EVP_sha1(),
				                  NULL);
			}
# else /* HAVE_SHA256 */
			(void) EVP_Digest(domain, strlen(domain), shaout, NULL,
			                  EVP_sha1(), NULL);
# endif /* HAVE_SHA256 */
#endif /* USE_GNUTLS */

			/* encode with base32 */
			memset(base32, '\0', sizeof base32);
			b32len = sizeof base32 - 1;
			(void) dkim_base32_encode(base32, &b32len, shaout,
			                          shalen);
		}

		/* generate output */
		snprintf(value, sizeof value, "%s; d=%s", VALIDATPS, domain);

		if (nsupdate)
		{
			fprintf(out, "update add %s%s TXT %d \"%s\"\n",
			        base32[0] == '\0' ? domain : base32,
			        suffix ? ATPSZONE : "",
				ttl == -1 ? defttl : ttl,
			        value);
		}
		else if (ttl == -1)
		{
			fprintf(out, "%s%s\tIN\tTXT\t\"%s\"\n",
			        base32[0] == '\0' ? domain : base32,
			        suffix ? ATPSZONE : "",
			        value);
		}
		else
		{
			fprintf(out, "%s%s\t%d\tIN\tTXT\t\"%s\"\n",
			        base32[0] == '\0' ? domain : base32,
			        suffix ? ATPSZONE : "", ttl,
			        value);
		}
	}

	(void) dkimf_db_close(db);

	if (nsupdate)
		fprintf(out, "send\nanswer\n");

	if (out != stdout)
		fclose(out);

	if (verbose > 0)
	{
		fprintf(stdout, "%s: %d record%s written\n",
		        progname, c, c == 1 ? "" : "s");
	}

	return 0;
}
